James-xxxx: FluentBit documentation (#3850)
linagora/james-project: docs/modules/servers/pages/distributed/operate/logging.adoc

review of performances at the tier level.

Leveraging filters in Kibana discover view can help filtering out
''already known'' frequently occurring logs.

When reporting ERROR or WARNING logs, consider adding the full logs, and
related data (eg the raw content of a mail triggering an issue) to the
bug report in order to ease resolution.

== Logging configuration

Distributed James uses link:http://logback.qos.ch/[logback] as a logging library
and link:https://docs.fluentbit.io/[FluentBit] as centralize logging.

Information about logback configuration can be found
link:http://logback.qos.ch/manual/configuration.html[here].

> == Structured logging

Reviewer: We should have the following structure:

```
== Structured logging

=== Pushing logs to ElasticSearch

[... What was done before ]

=== Using fluentBit as a log forwarder

[ The new stuff ]

```
```

                <encoder class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
                    <layout class="ch.qos.logback.contrib.json.classic.JsonLayout">
                        <timestampFormat>yyyy-MM-dd'T'HH:mm:ss.SSSX</timestampFormat>
                        <timestampFormatTimezoneId>Etc/UTC</timestampFormatTimezoneId>

                        <!-- Importance for handling multiple lines log -->
                        <appendLineSeparator>true</appendLineSeparator>

                        <jsonFormatter class="ch.qos.logback.contrib.jackson.JacksonJsonFormatter">
                            <prettyPrint>false</prettyPrint>
                        </jsonFormatter>
                    </layout>
                </encoder>
        </appender>

        <root level="INFO"><appender-ref ref="LOG_FILE" />
        </root>

        <root level="DEBUG"><appender-ref ref="LOG_FILE" />
        </root>

        <root level="WARN">
                <appender-ref ref="LOG_FILE" />
        </root>

        <logger name="org.apache.james" level="INFO" />
        <logger name="com.datastax.driver.core.QueryLogger.SLOW" level="DEBUG" />

</configuration>

> ....

Reviewer: Please also share the fluentBit configuration you used

Here is a sample `conf/logback.xml` configuration file for logback with the following
pre-requisites:

> * Logging both in an unstructured fashion on the console and in a structure fashion in ElasticSearch

Reviewer 2: ```suggestion
* Logging both in an unstructured fashion on the console and in a structured fashion in ElasticSearch
```

> * Logging ElasticSearch Log appender logs in the console

Reviewer: We no longer dispay the sample logback.xml file for ElasticSearch exporter?

> === Using fluentBit as a log forwarder

Reviewer 2: ```suggestion
=== Using FluentBit as a log forwarder
```

> Distributed Server leverages the use of MDC in order to achieve structured logging, and better add context to the logged information. We furthermore ship json log in to file with RollingFileAppender on the classpath to easily allow FluentBit direct tail the log file.

Reviewer 2: ```suggestion
Distributed Server leverages the use of MDC in order to achieve structured logging, and better add context to the logged information. We furthermore ship json logs to file with RollingFileAppender on the classpath to easily allow FluentBit to directly tail the log file.
```

Here is a sample conf/logback.xml configuration file for logback with the following pre-requisites:

> Logging in a structure json fashion and write to file for centralize logging.

Reviewer 2: ```suggestion
Logging in a structured json fashion and write to file for centralizing logging.
```

> Centralize logging third party like FluentBit can tail from logging’s file then filter/process and put in to ElastichSearch

Reviewer 2: ```suggestion
To centralize logging, a third party software like FluentBit can tail from logging’s file then filter/process and put in to ElastichSearch
```

....
<?xml version="1.0" encoding="UTF-8"?>
<configuration>

        <contextListener class="ch.qos.logback.classic.jul.LevelChangePropagator">
                <resetJUL>true</resetJUL>
        </contextListener>

>         <appender name="LOG_FILE" class="ch.qos.logback.core.rolling.RollingFileAppender">

Reviewer: Can we do that with the Console Appender instead ?

                    <layout class="ch.qos.logback.contrib.json.classic.JsonLayout">
                        <timestampFormat>yyyy-MM-dd'T'HH:mm:ss.SSSX</timestampFormat>
                        <timestampFormatTimezoneId>Etc/UTC</timestampFormatTimezoneId>

                        <!-- Importance for handling multiple lines log -->
                        <appendLineSeparator>true</appendLineSeparator>

                        <jsonFormatter class="ch.qos.logback.contrib.jackson.JacksonJsonFormatter">
                            <prettyPrint>false</prettyPrint>
                        </jsonFormatter>
                    </layout>
                </encoder>
        </appender>

        <root level="INFO"><appender-ref ref="LOG_FILE" />
        </root>

        <root level="DEBUG"><appender-ref ref="LOG_FILE" />
        </root>

        <root level="WARN">
                <appender-ref ref="LOG_FILE" />
        </root>

        <logger name="org.apache.james" level="INFO" />
        <logger name="com.datastax.driver.core.QueryLogger.SLOW" level="DEBUG" />

</configuration>
....

First you need to create a logs folder to mount the `logs` folder of James outside container,

> then mount that folder to FluentBit. In case you need to use `sh command` in fluent-bit container please use image: `fluent/fluent-bit:1.5.7-debug`

Reviewer 2: First you need to create a `logs` folder, then mount it to James container and to FluentBit.

      - "993:993"
      - "8080:8000"

  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch:6.3.2
    ports:
      - "9200:9200"
    environment:
      - discovery.type=single-node

  cassandra:
    image: cassandra:3.11.3
    ports:
      - "9042:9042"

  rabbitmq:
    image: rabbitmq:3.8.1-management
    ports:
      - "5672:5672"
      - "15672:15672"

  s3:
    image: zenko/cloudserver:8.2.6
    container_name: s3.docker.test
    environment:
      - SCALITY_ACCESS_KEY_ID=accessKey1
      - SCALITY_SECRET_ACCESS_KEY=secretKey1
      - S3BACKEND=mem
      - LOG_LEVEL=trace
      - REMOTE_MANAGEMENT_DISABLE=1

  fluent-bit:
    image: fluent/fluent-bit:1.5.7
    volumes:
      - ./fluentbit/fluent-bit.conf:/fluent-bit/etc/fluent-bit.conf
      - ./fluentbit/parsers.conf:/fluent-bit/etc/parsers.conf
      - ./logs:/fluent-bit/log
    ports:
      - "24224:24224"
      - "24224:24224/udp"
    depends_on:
      - elasticsearch

  kibana:
    image: docker.elastic.co/kibana/kibana:6.3.2
    environment:
      ELASTICSEARCH_HOSTS: http://elasticsearch:9200
    ports:
      - "5601:5601"
    depends_on:
      - elasticsearch
....

FluentBit config as:

> the `Host elasticsearch` it point to `elasticsearch` service in docker-compose file.

Reviewer 2: ```suggestion
the `Host elasticsearch` pointing to `elasticsearch` service in docker-compose file.
```

....
[SERVICE]
    Parsers_File    /fluent-bit/etc/parsers.conf

[INPUT]
    name                    tail
    path                    /fluent-bit/log/*.log
    Parser                  docker
    docker_mode             on
    buffer_chunk_size       1MB
    buffer_max_size         1MB
    mem_buf_limit           64MB
    Refresh_Interval        30

[OUTPUT]
    Name  stdout
    Match *

[OUTPUT]
    Name  es
    Match *
    Host elasticsearch
    Port 9200
    Index fluentbit
    Logstash_Format On
    Logstash_Prefix fluentbit-james
    Type docker
....

> FulentBit Parser config:

Reviewer 2: ```suggestion
FluentBit Parser config:
```
